### ISSUE TITLE
[analyzer] Unexpected `non_exhaustive_switch_expression` error

### ISSUE THREAD

**Issue by incendial** [dart-lang/sdk#52782]
Consider the following case:

```dart
void main() {  
  SomeClass? first = SomeClass();
  SomeClass? second = SomeClass();
  
  final result = switch ((first, second)) { // <-- shows an error
    (== null, != null) => true,
    (!= null, == null) => false,
    (== null, == null) => false,
    (!= null, != null) => process(first, second),
  };
}

bool process(SomeClass first, SomeClass second) => true;

class SomeClass {}
```

Analyzer shows: The type '(SomeClass, SomeClass)' is not exhaustively matched by the switch cases since it doesn't match '(SomeClass(), SomeClass())', but should the last line (`(!= null, != null)`) cover this case?

**Comment by lrhn** [comment]
Agree it's probably an analyzer bug.
CFE accepts the code, and it looks very much like something that should work.

(The example is even more weird because the matched value type is `(SomeClass, SomeClass)`, so the first three cases are statically known to not match, and the fourth case should, equally statically, be known to always match.
It requires unsound null safety, which no longer exists, but may still be in the tool code, to have the fourth case not match, but then one of the first three still still should.)

**Comment by incendial** [comment]
Yeah, great point. Even if I remove the `?` from declarations, there is still no error about the first three matches...

**Comment by bwilkerson** [comment]
@scheglov

**Comment by scheglov** [comment]
I believe this implementation is  from `_fe_analyzer_shared`. @johnniwinther 
CFE reports similar error.
```
bin/test.dart:3:19: Error: The type 'Object?' is not exhaustively matched by the switch cases since it doesn't match 'null'.
 - 'Object' is from 'dart:core'.
Try adding a wildcard pattern or cases that match 'null'.
  var v = switch (o) {
                  ^
```
